src/V3Randomize.cpp

```diff
@@ -1323,23 +1323,77 @@ class RandomizeVisitor final : public VNVisitor {
-                    stmtsp->addNext(randp);
-                } else {
-                    stmtsp = randp;
-                }
+                stmtsp = stmtsp ? stmtsp->addNext(randp) : randp;
             }
             return stmtsp;
-        } else if (const auto* const unionDtp = VN_CAST(memberp ? memberp->subDTypep()->skipRefp()
-                                                                : exprp->dtypep()->skipRefp(),
-                                                        UnionDType)) {
+        } else if (const auto* const unionDtp = VN_CAST(memberDtp, UnionDType)) {
             if (!unionDtp->packed()) {
                 unionDtp->v3error("Unpacked unions shall not be declared as rand or randc."
                                   " (IEEE 1800-2023 18.4)");
                 return nullptr;
             }
             AstMemberDType* const firstMemberp = unionDtp->membersp();
             return newRandStmtsp(fl, exprp, nullptr, offset, firstMemberp);
+        } else if (AstDynArrayDType* const dynarrayDtp = VN_CAST(memberDtp, DynArrayDType)) {
+            V3UniqueNames* uniqueNamep = new V3UniqueNames{"__Vrandtemp"};
+            AstNodeDType* tempDTypep = dynarrayDtp;
+            AstCMethodHard* tempDynArrayMethodp = nullptr;
+            AstVar* randLoopIndxp = nullptr;
+            while (VN_CAST(tempDTypep, DynArrayDType)) {
+
+                AstVar* tempRandLoopIndxp
+                    = new AstVar{fl, VVarType::VAR, uniqueNamep->get(""),
+                                 dynarrayDtp->findBasicDType(VBasicDTypeKwd::UINT32)};
+
+                if (randLoopIndxp) {
+                    randLoopIndxp->addNext(tempRandLoopIndxp);
+                } else {
+                    randLoopIndxp = tempRandLoopIndxp;
+                }
+                tempDynArrayMethodp
+                    = tempDynArrayMethodp
+                          ? new AstCMethodHard{fl, tempDynArrayMethodp, "atWrite",
+                                               new AstVarRef{fl, tempRandLoopIndxp, VAccess::READ}}
+                          : new AstCMethodHard{
+                              fl, exprp, "atWrite",
+                              new AstVarRef{fl, tempRandLoopIndxp, VAccess::READ}};
+                tempDynArrayMethodp->dtypep(tempDTypep->subDTypep());
+                tempDTypep = tempDTypep->virtRefDTypep();
+            }
+            AstSelLoopVars* randLoopVarp = new AstSelLoopVars{
+                fl, new AstVarRef{fl, ((AstVarRef*)exprp)->varp(), VAccess::READ}, randLoopIndxp};
+            AstNodeStmt* stmtsp = new AstForeach{fl, randLoopVarp,
+                                                 newRandStmtsp(fl, tempDynArrayMethodp, nullptr)};
+            return stmtsp;
+        } else if (AstUnpackArrayDType* const unpackarrayDtp
+                   = VN_CAST(memberDtp, UnpackArrayDType)) {
+            V3UniqueNames* randUniqueVarp = new V3UniqueNames{"__Vrandtemp"};
+            AstNodeDType* tempDTypep = unpackarrayDtp;
```

Comment:
Similar changes in this code to those suggested above.  Perhaps this is similar enough it can be a common function between both?